pkg/controller/restic_repository_controller.go

```diff
@@ -0,0 +1,276 @@
+	c.logger.Debug("resticRepositoryController.enqueueAllRepositories")
+
+	repos, err := c.resticRepositoryLister.List(labels.Everything())
+	if err != nil {
+		c.logger.WithError(errors.WithStack(err)).Error("error listing restic repositories")
+		return
+	}
+
+	for _, repo := range repos {
+		c.enqueue(repo)
+	}
+}
+
+func (c *resticRepositoryController) processQueueItem(key string) error {
+	log := c.logger.WithField("key", key)
+	log.Debug("Running processQueueItem")
+
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		log.WithError(errors.WithStack(err)).Error("error splitting queue key")
+		return nil
+	}
+
+	log = c.logger.WithField("namespace", ns).WithField("name", name)
+
+	req, err := c.resticRepositoryLister.ResticRepositories(ns).Get(name)
+	if apierrors.IsNotFound(err) {
+		log.Debug("Unable to find ResticRepository")
+		return nil
+	}
+	if err != nil {
+		return errors.Wrap(err, "error getting ResticRepository")
+	}
+
+	// Don't mutate the shared cache
+	reqCopy := req.DeepCopy()
+
+	switch req.Status.Phase {
+	case "", v1.ResticRepositoryPhaseNew:
+		return c.initializeRepo(reqCopy, log)
+	case v1.ResticRepositoryPhaseReady:
+		return c.runMaintenanceIfDue(reqCopy, log)
+	case v1.ResticRepositoryPhaseNotReady:
+		return c.checkNotReadyRepo(reqCopy, log)
+	}
+
+	return nil
+}
+
+func (c *resticRepositoryController) initializeRepo(req *v1.ResticRepository, log logrus.FieldLogger) error {
+	log.Info("Initializing restic repository")
+
+	// defaulting - if the patch fails, return an error so the item is returned to the queue
+	if err := c.patchResticRepository(req, func(r *v1.ResticRepository) {
+		r.Spec.ResticIdentifier = restic.GetRepoIdentifier(c.objectStorageConfig, r.Name)
+
+		if r.Spec.MaintenanceFrequency.Duration <= 0 {
+			r.Spec.MaintenanceFrequency = metav1.Duration{Duration: restic.DefaultMaintenanceFrequency}
+		}
+	}); err != nil {
+		return err
+	}
+
+	if err := ensureRepo(req.Name, c.repositoryManager); err != nil {
+		return c.patchResticRepository(req, repoNotReady(err.Error()))
+	}
+
+	return c.patchResticRepository(req, func(req *v1.ResticRepository) {
+		req.Status.Phase = v1.ResticRepositoryPhaseReady
+		req.Status.LastMaintenanceTime = metav1.Time{Time: time.Now()}
+	})
+}
+
+// ensureRepo first checks the repo, and returns if check passes. If it fails,
+// attempts to init the repo, and returns the result.
+func ensureRepo(name string, repoManager restic.RepositoryManager) error {
+	if repoManager.CheckRepo(name) == nil {
+		return nil
+	}
+
+	return repoManager.InitRepo(name)
+}
+
+func (c *resticRepositoryController) runMaintenanceIfDue(req *v1.ResticRepository, log logrus.FieldLogger) error {
+	log.Debug("resticRepositoryController.runMaintenanceIfDue")
+
+	now := c.clock.Now()
+
+	if !dueForMaintenance(req, now) {
+		log.Debug("not due for maintenance")
+		return nil
+	}
+
+	log.Info("Running maintenance on restic repository")
+
+	log.Debug("Checking repo before prune")
```

Comment:
Launching an Ark server locally with the `resticLocation` removed from the config, but a `ResticRepository` existing in a namespace caused a panic, with this being the last log message:

```
time="2018-06-19T15:17:01-04:00" level=info msg="setting log-level to DEBUG"
time="2018-06-19T15:17:01-04:00" level=info msg="Starting Ark server v0.9.0-alpha.2-16-g332af0a-dirty" logSource="pkg/cmd/server/server.go:100"
time="2018-06-19T15:17:01-04:00" level=info msg="Ensuring namespace exists for backups" logSource="pkg/cmd/server/server.go:258" namespace=heptio-ark
time="2018-06-19T15:17:01-04:00" level=info msg="Namespace already exists" logSource="pkg/cmd/server/server.go:270" namespace=heptio-ark
time="2018-06-19T15:17:01-04:00" level=info msg="Retrieving Ark configuration" logSource="pkg/cmd/server/server.go:275"
time="2018-06-19T15:17:01-04:00" level=info msg="Successfully retrieved Ark configuration" logSource="pkg/cmd/server/server.go:293"
time="2018-06-19T15:17:01-04:00" level=info msg="Using default resource priorities" logSource="pkg/cmd/server/server.go:343" priorities="[namespaces persistentvolumes persistentvolumeclaims secrets configmaps serviceaccounts limitranges pods]"
time="2018-06-19T15:17:01-04:00" level=info msg="Configuring cloud provider for backup service" logSource="pkg/cmd/server/server.go:392"
time="2018-06-19T15:17:01-04:00" level=debug msg="starting plugin" args="[./_output/bin/linux/amd64/ark run-plugin cloudprovider aws]" logSource="pkg/plugin/logrus_adapter.go:74" path=./_output/bin/linux/amd64/ark
time="2018-06-19T15:17:01-04:00" level=debug msg="waiting for RPC address" logSource="pkg/plugin/logrus_adapter.go:74" path=./_output/bin/linux/amd64/ark
time="2018-06-19T15:17:01-04:00" level=debug msg="plugin address" address=/tmp/plugin894367372 logSource="pkg/plugin/logrus_adapter.go:74" network=unix pluginName=ark
time="2018-06-19T15:17:01-04:00" level=info msg="Configuring cloud provider for snapshot service" logSource="pkg/cmd/server/server.go:409"
time="2018-06-19T15:17:01-04:00" level=info msg="Starting controllers" logSource="pkg/cmd/server/server.go:495"
time="2018-06-19T15:17:01-04:00" level=info msg="Caching cloud backups every 30m0s" logSource="pkg/cmd/server/server.go:501"
time="2018-06-19T15:17:01-04:00" level=debug msg="refreshing all cached backup lists from object storage" logSource="pkg/cloudprovider/backup_cache.go:69"
time="2018-06-19T15:17:01-04:00" level=info msg="Starting metric server at address [:8085]" logSource="pkg/cmd/server/server.go:512"
time="2018-06-19T15:17:01-04:00" level=info msg="Running backup sync controller" logSource="pkg/controller/backup_sync_controller.go:70"
time="2018-06-19T15:17:01-04:00" level=info msg="Syncing backups from object storage" logSource="pkg/controller/backup_sync_controller.go:78"
time="2018-06-19T15:17:01-04:00" level=debug msg="Bucket is not in cache - doing a live lookup" bucket=<REDACTED> logSource="pkg/cloudprovider/backup_cache.go:89"
time="2018-06-19T15:17:02-04:00" level=info msg="Got backups from object storage" backupCount=1 logSource="pkg/controller/backup_sync_controller.go:84"
time="2018-06-19T15:17:02-04:00" level=info msg="Syncing backup" backup=heptio-ark/nginx-logs logSource="pkg/controller/backup_sync_controller.go:88"
time="2018-06-19T15:17:06-04:00" level=info msg="Starting BackupController" logSource="pkg/controller/backup_controller.go:153"
time="2018-06-19T15:17:06-04:00" level=info msg="Starting ScheduleController" logSource="pkg/controller/schedule_controller.go:135"
time="2018-06-19T15:17:06-04:00" level=info msg="Waiting for caches to sync" logSource="pkg/controller/backup_controller.go:156"
time="2018-06-19T15:17:06-04:00" level=info msg="Waiting for caches to sync" logSource="pkg/controller/schedule_controller.go:138"
time="2018-06-19T15:17:06-04:00" level=info msg="Starting controller" controller=backup-deletion logSource="pkg/controller/generic_controller.go:77"
time="2018-06-19T15:17:06-04:00" level=info msg="Starting controller" controller=gc-controller logSource="pkg/controller/generic_controller.go:77"
time="2018-06-19T15:17:06-04:00" level=info msg="Waiting for caches to sync" controller=backup-deletion logSource="pkg/controller/generic_controller.go:80"
time="2018-06-19T15:17:06-04:00" level=info msg="Waiting for caches to sync" controller=gc-controller logSource="pkg/controller/generic_controller.go:80"
time="2018-06-19T15:17:06-04:00" level=info msg="Starting RestoreController" logSource="pkg/controller/restore_controller.go:154"
time="2018-06-19T15:17:06-04:00" level=info msg="Starting DownloadRequestController" logSource="pkg/controller/download_request_controller.go:122"
time="2018-06-19T15:17:06-04:00" level=info msg="Waiting for caches to sync" logSource="pkg/controller/restore_controller.go:157"
time="2018-06-19T15:17:06-04:00" level=info msg="Starting controller" controller=restic-repository logSource="pkg/controller/generic_controller.go:77"
time="2018-06-19T15:17:06-04:00" level=info msg="Waiting for caches to sync" logSource="pkg/controller/download_request_controller.go:125"
time="2018-06-19T15:17:06-04:00" level=info msg="Waiting for caches to sync" controller=restic-repository logSource="pkg/controller/generic_controller.go:80"
time="2018-06-19T15:17:06-04:00" level=info msg="Caches are synced" controller=restic-repository logSource="pkg/controller/generic_controller.go:84"
time="2018-06-19T15:17:06-04:00" level=debug msg=resticRepositoryController.enqueueAllRepositories controller=restic-repository logSource="pkg/controller/restic_repository_controller.go:88"
time="2018-06-19T15:17:06-04:00" level=debug msg="Running processQueueItem" controller=restic-repository key=heptio-ark/nginx-example logSource="pkg/controller/restic_repository_controller.go:103"
time="2018-06-19T15:17:06-04:00" level=debug msg=resticRepositoryController.runMaintenanceIfDue controller=restic-repository logSource="pkg/controller/restic_repository_controller.go:172" name=nginx-example namespace=heptio-ark
time="2018-06-19T15:17:06-04:00" level=info msg="Running maintenance on restic repository" controller=restic-repository logSource="pkg/controller/restic_repository_controller.go:181" name=nginx-example namespace=heptio-ark
time="2018-06-19T15:17:06-04:00" level=debug msg="Checking repo before prune" controller=restic-repository logSource="pkg/controller/restic_repository_controller.go:183" name=nginx-example namespace=heptio-ark
E0619 15:18:10.870497    7295 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:388
/go/src/github.com/heptio/ark/pkg/controller/restic_repository_controller.go:184
/go/src/github.com/heptio/ark/pkg/controller/restic_repository_controller.go:129
/go/src/github.com/heptio/ark/pkg/controller/restic_repository_controller.go:70
/go/src/github.com/heptio/ark/pkg/controller/generic_controller.go:128
/go/src/github.com/heptio/ark/pkg/controller/generic_controller.go:115
/go/src/github.com/heptio/ark/pkg/controller/generic_controller.go:89
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/go/src/github.com/heptio/ark/pkg/controller/generic_controller.go:89
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x10d79bc]

goroutine 241 [running]:
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x17f33c0, 0x27eb5b0)
        /usr/local/go/src/runtime/panic.go:502 +0x229
github.com/heptio/ark/pkg/controller.(*resticRepositoryController).runMaintenanceIfDue(0xc4200fd380, 0xc42027b8c0, 0x1c6dee0, 0xc420592730, 0x0, 0x0)
        /go/src/github.com/heptio/ark/pkg/controller/restic_repository_controller.go:184 +0x1dc
github.com/heptio/ark/pkg/controller.(*resticRepositoryController).processQueueItem(0xc4200fd380, 0xc42003edc0, 0x18, 0x18, 0xc4207a0db8)
        /go/src/github.com/heptio/ark/pkg/controller/restic_repository_controller.go:129 +0x4cb
github.com/heptio/ark/pkg/controller.(*resticRepositoryController).(github.com/heptio/ark/pkg/controller.processQueueItem)-fm(0xc42003edc0, 0x18, 0xc420205f40, 0x16d6080)
        /go/src/github.com/heptio/ark/pkg/controller/restic_repository_controller.go:70 +0x3e
github.com/heptio/ark/pkg/controller.(*genericController).processNextWorkItem(0xc420091200, 0xc4200e1b00)
        /go/src/github.com/heptio/ark/pkg/controller/generic_controller.go:128 +0xd1
github.com/heptio/ark/pkg/controller.(*genericController).runWorker(0xc420091200)
        /go/src/github.com/heptio/ark/pkg/controller/generic_controller.go:115 +0x2b
github.com/heptio/ark/pkg/controller.(*genericController).(github.com/heptio/ark/pkg/controller.runWorker)-fm()
        /go/src/github.com/heptio/ark/pkg/controller/generic_controller.go:89 +0x2a
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc4205857a0)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4206cffa0, 0x3b9aca00, 0x0, 0xc420585601, 0xc42004e180)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc4205857a0, 0x3b9aca00, 0xc42004e180)
        /go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
github.com/heptio/ark/pkg/controller.(*genericController).Run.func2(0xc420091200, 0x1c51620, 0xc4204b9340, 0xc420579510)
        /go/src/github.com/heptio/ark/pkg/controller/generic_controller.go:89 +0x68
created by github.com/heptio/ark/pkg/controller.(*genericController).Run
        /go/src/github.com/heptio/ark/pkg/controller/generic_controller.go:88 +0x2f6
```

Reply:
Pushed an update to not run the controller if no `resticLocation` is configured.